kafka/event.go

```diff
@@ -145,170 +147,179 @@ func (o OAuthBearerTokenRefresh) String() string {
+
+			return peof, nil
+
+		} else if int(C.rd_kafka_event_error_is_fatal(rkev)) != 0 {
+			// A fatal error has been raised.
+			// Extract the actual error from the client
+			// instance and return a new Error with
+			// fatal set to true.
+			cFatalErrstrSize := C.size_t(512)
+			cFatalErrstr := (*C.char)(C.malloc(cFatalErrstrSize))
+			defer C.free(unsafe.Pointer(cFatalErrstr))
+			cFatalErr := C.rd_kafka_fatal_error(h.rk, cFatalErrstr, cFatalErrstrSize)
+			fatalErr := newErrorFromCString(cFatalErr, cFatalErrstr)
+			fatalErr.fatal = true
+			return fatalErr, nil
+
+		} else {
+			return newErrorFromCString(cErr, C.rd_kafka_event_error_string(rkev)), nil
+		}
 
-				} else {
-					retval = msg
-					break out
-				}
-			}
+	case C.RD_KAFKA_EVENT_STATS:
+		return &Stats{C.GoString(C.rd_kafka_event_stats(rkev))}, nil
 
-		case C.RD_KAFKA_EVENT_OFFSET_COMMIT:
-			// Offsets committed
-			cErr := C.rd_kafka_event_error(rkev)
-			coffsets := C.rd_kafka_event_topic_partition_list(rkev)
-			var offsets []TopicPartition
-			if coffsets != nil {
-				offsets = newTopicPartitionsFromCparts(coffsets)
-			}
+	case C.RD_KAFKA_EVENT_DR:
+		// Producer Delivery Report event
+		// Each such event contains delivery reports for all
+		// messages in the produced batch.
+		// Forward delivery reports to per-message's response channel
+		// or to the global Producer.Events channel, or none.
+		rkmessages := make([]*C.rd_kafka_message_t, int(C.rd_kafka_event_message_count(rkev)))
 
-			if cErr != C.RD_KAFKA_RESP_ERR_NO_ERROR {
-				retval = OffsetsCommitted{newErrorFromCString(cErr, C.rd_kafka_event_error_string(rkev)), offsets}
-			} else {
-				retval = OffsetsCommitted{nil, offsets}
-			}
+		cnt := int(C.rd_kafka_event_message_array(rkev, (**C.rd_kafka_message_t)(unsafe.Pointer(&rkmessages[0])), C.size_t(len(rkmessages))))
 
-		case C.RD_KAFKA_EVENT_OAUTHBEARER_TOKEN_REFRESH:
-			ev := OAuthBearerTokenRefresh{C.GoString(C.rd_kafka_event_config_string(rkev))}
-			retval = ev
+		for _, rkmessage := range rkmessages[:cnt] {
+			msg := h.newMessageFromC(rkmessage)
+			var ch *chan Event
 
-		case C.RD_KAFKA_EVENT_NONE:
-			// poll timed out: no events available
-			break out
+			if rkmessage._private != nil {
+				// Find cgoif by id
+				cg, found := h.cgoGet((int)((uintptr)(rkmessage._private)))
+				if found {
+					cdr := cg.(cgoDr)
 
-		default:
-			if rkev != nil {
-				fmt.Fprintf(os.Stderr, "Ignored event %s\n",
-					C.GoString(C.rd_kafka_event_name(rkev)))
+					if cdr.deliveryChan != nil {
+						ch = &cdr.deliveryChan
+					}
+					msg.Opaque = cdr.opaque
+				}
 			}
 
-		}
-
-		if retval != nil {
-			if channel != nil {
+			if ch != nil {
 				select {
-				case channel <- retval:
-				case <-termChan:
-					retval = nil
-					term = true
-					break out
+				case *ch <- msg:
+				case <-ctx.Done():
+					// This is a _very_ dangerous codepath. If the context passed in is canceled, and causes us
```

Comment:
I'm not comfortable with this, all the delivery reports in this event should be forwarded before checking the context.
As for the deadlock/unserved channel on termination; that is an existing problem and not something that needs to be fixed now, and there are proper workarounds (call Flush() and read the dr chan in separate go-routines).

Reply:
I need to have a long think about this in front of the whiteboard with @seancaffery I think - I suspect my understanding of how the shutdown process is supposed to work might be wrong.

Once I've done that, yeah, I'll probably remove this check.

Reply:
OK - I think we've nutted out how the shutdown is supposed to work here. Agree that we should block on `*ch <- msg` irrespective of the context being canceled or not. If this behaviour is _not_ what a user of the library wants, they can pass in a buffered channel to `Produce` - then this line will never block regardless of what the application is doing (and the DR's will be thrown in the bin just like the above code is doing now).